src/main/java/de/unistuttgart/iste/gits/user_service/service/MembershipService.java

```diff
@@ -110,6 +112,39 @@ public CourseMembership deleteMembership(CourseMembershipInput inputDto) {
         return membershipMapper.entityToDto(entity);
     }
 
+    /**
+     * removes all memberships for a given course ID from the database
+     * @param courseId valid course ID
+     */
+    public void deleteMembershipByCourseId(UUID courseId){
+
+        List<CourseMembershipEntity> memberships = courseMembershipRepository.findCourseMembershipEntitiesByCourseId(courseId);
+
+        if (memberships != null && !memberships.isEmpty()){
+            courseMembershipRepository.deleteAll(memberships);
+        }
+    }
+
+    /**
+     * method to handle course deletion events. removes all course memberships associated with the course ID
+     * @param changeEvent course change event
+     */
+    public void removeCourse(CourseChangeEvent changeEvent){
+        // evaluate course Update message
+        if (changeEvent.getCourseId() == null || changeEvent.getOperation() == null){
+            throw new NullPointerException("incomplete message received: all fields of a message must be non-null");
```

Comment:
Actually this is dangerous. When dapr calls a service with an event and this service returns 500 (or any other code other than 2XX), dapr will continue to call this service with the event (at-least-once semantics). So if for some reason, null is one of these fields, this will make dapr call this service with the malformed event until the end of time. (This is what currently happens in the reward service and it's a mess)

Here in the service it makes sense to throw an exception, but I would wrap the call in the subscription controller with try catch and catching that exception. Maybe change it to a "ValidationException", so not any null pointer exception is catched. 

Just my suggestion, maybe you have a better idea how to deal with it.